### PR TITLE
GrapesJS with MJML Version 4

### PR DESCRIPTION
**Pull request by larsdecker**
This version runs with the MJML Version 4 and has only a few bugs. Maybe it is a good idea to put into a separate dev branch.

See Issue here: https://github.com/artf/grapesjs-mjml/issues/73

### CODE REVIEW

**Comment by larsdecker** [comment]
@artf 
It would be also be good, to put a dev versions into the npm registry.

**Comment by CimattiConsulting** [comment]
@larsdecker 
The insertion of the columns has stopped working after your changes

**Comment by larsdecker** [comment]
@CimattiConsulting 
Have you got an example for me?

I tested it several times and it seems to be working.

**Comment by CimattiConsulting** [comment]
@larsdecker 
My fault, I was trying another version.

**Comment by CimattiConsulting** [comment]
Currently i'm working on the width problem of the mj-body

**Comment by larsdecker** [comment]
Okay, that is good. I have added yesterday the undo / redo functions and the device icons from the Newsletter-Preset to this project as a new branch on my repo.

**Comment by CimattiConsulting** [comment]
> @larsdecker
> My fault, I was trying another version.

@larsdecker 
Try adding a column after changing the width of the mj-body

**Comment by larsdecker** [comment]
> > @CimattiConsulting
> > Have you got an example for me?
> > I tested it several times and it seems to be working.
> 
> @larsdecker
> Try adding a column after changing the width of the mj-body

Yep, then it is broken :-(

**Comment by CimattiConsulting** [comment]
> Okay, that is good. I have added yesterday the undo / redo functions and the device icons from the Newsletter-Preset to this project as a new branch on my repo.

@larsdecker 
I don't think it was a good idea to copy the code from the Preset-newsletter without checking it.
Many settings have nothing to do with grapesjs-mjml

**Comment by larsdecker** [comment]
> > Okay, that is good. I have added yesterday the undo / redo functions and the device icons from the Newsletter-Preset to this project as a new branch on my repo.
> 
> @larsdecker
> I don't think it was a good idea to copy the code from the Preset-newsletter without checking it.
> Many settings have nothing to do with grapesjs-mjml

I agree with you that it was not a good idea to adopt all the settings. But for the rest of the code, I think it's okay. I will check the settings and will only adopt these, that are actually is using.

**Comment by larsdecker** [comment]
I aggreed for the Settings. This was not a good idea. I will check the settings which are actually use in the grapesjs-mjml. For the rest of the code, i think it was okay.

**Comment by CimattiConsulting** [comment]
>
> @larsdecker
> Try adding a column after changing the width of the mj-body

I looked at it but I didn't find anything for now, can you check it too?

**Comment by larsdecker** [comment]
I have also looked at it, but i find also not a solution for it. It directly breaks when the width was changed. For the time being we can remove the possibility to change the width.

**Comment by CimattiConsulting** [comment]
Wait... 
it didn't even work in the old version.
I don't think it's a good idea to remove the width attribute also because it's a problem of all the attributes of the mj-body

**Comment by CimattiConsulting** [comment]
Fixed it!!!
-> larsdecker/grapesjs-mjml#4

**Comment by larsdecker** [comment]
Very good job. Thanks a lot @CimattiConsulting

**Comment by CimattiConsulting** [comment]
@larsdecker @artf 
I think we're ready for merge

**Comment by larsdecker** [comment]
Yes, i think we are ready for a merge into the master branch and a new release into npm.

@artf 
Please test it and merge it.

**Comment by DRoet** [comment]
Great effort in this PR guys, maybe we can PR against `mjml` to make it work in the browser instead of depending on `mjml4-in-browser`? see https://github.com/mjmlio/mjml/issues/928#issuecomment-508189319

Also shouldn't `mjml4-in-browser` be a replacement of `mjml: ^4.0.0`? it currently pulls in both in the `package.json` file

**Comment by larsdecker** [comment]
@DRoet 
Yes indeed, at the moment mjml is loaded into two folders. But the real mjml dependency is not used.
 I didn´t know that before. Anyone has an idea, how we can PR a fix to mjml that it work in the browser?

**Comment by CimattiConsulting** [comment]
@bures
Do you have any idea?

**Comment by CimattiConsulting** [comment]
@larsdecker 
maybe, we could remove mjml-in-browser from dependencies and replicate it on grapesjs-mjml

**Comment by larsdecker** [comment]
@CimattiConsulting  
I try it, but i ran into some issues.

```WARNING in ./node_modules/mjml-core/lib/helpers/mjmlconfig.js
109:27-48 Critical dependency: the request of a dependency is an expression
 @ ./node_modules/mjml-core/lib/helpers/mjmlconfig.js
 @ ./node_modules/mjml-core/lib/index.js
 @ ./node_modules/mjml/lib/index.js
 @ ./src/components/index.js
 @ ./src/index.js
 @ multi (webpack)-dev-server/client?http://localhost:8080 ./src/index
```

I have not enough experiences with webpack, to fix that issue.

File: build/config.gypi [hunk 1]
```diff
@@ -0,0 +1,72 @@
+# Do not edit. File was generated by node-gyp's "configure" step
```
Comment (artf) [inline build/config.gypi]: Do we need this file? 🤔

Comment (larsdecker) [reply build/config.gypi]: I have delete that file. I didn´t know what is it.

**Comment by bures** [comment]
I'm a bit lost in the conversation. mjml4-in-browser should be a replacement for mjml4. If I understand correctly, mjml4 should not be needed if you use mjml4-in-browser.

**Comment by CimattiConsulting** [comment]
@bures 
Yes, the problem is that grapesjs-mjml becomes dependent on mjml-in-browser for mjml updates

**Comment by bures** [comment]
Yep. I'm afraid that's the case. mjml4-in-browser cannot easily rely just on mjml4 package because it has to repackage the sources. If you like, I can make you a developer for mjml4-in-browser, so that you have the right to bump up the version of mjml4 dependency in mjml4-in-browser.

**Comment by CimattiConsulting** [comment]
@bures 
For the moment it may be fine, but it certainly cannot be considered a definitive solution.

**Comment by bures** [comment]
Sure. Propose anything better.

**Comment by larsdecker** [comment]
Yes. we should look at the original code of mjml and find a solution how it can be run in the browser.

**Comment by CimattiConsulting** [comment]
@larsdecker 
larsdecker/grapesjs-mjml#5

**Comment by bures** [comment]
My understanding is that the guys behind mjml don't care about running it in browser. That's why I created the mjml4-in-browser. It is the minimalistic wrapper around mjml4 that does not change anything from mjml4 only mocks things that do not work in browser.

**Comment by CimattiConsulting** [comment]
@bures 
Maybe they don't want to spend time on it,
but if we take care of making changes with all the people who need them,
I don't think they will refuse the pull request

**Comment by CimattiConsulting** [comment]
@artf 
Now it's okay?

**Comment by larsdecker** [comment]
@bures 
Can you please update the npm version of mjml-in-browser?

**Comment by bures** [comment]
Done. Please note that I exported some additional symbols, so that one can create custom components, etc. That's why I bumped the version to 1.1.2. Please check that it still works.

**Comment by larsdecker** [comment]
Thank you. Seems work fine in the editor.

**Comment by bures** [comment]
Excellent. Thanks.

**Comment by artf** [comment]
Thanks to everyone involved in this PR, especially for those who tried to understand undocumented stuff 😬 

I'll try to publish a new release ASAP

**Comment by CimattiConsulting** [comment]
@artf 
Can you set my private account ( @michibat01 ) as a collaborator, please?

**Comment by DRoet** [comment]
@artf I'm a bit late to the party, but could you add me as a collaborator aswell?

**Pull request merged**